src/v/storage/disk_log_impl.cc

```diff
@@ -2516,9 +2522,22 @@ disk_log_impl::remove_prefix_full_segments(truncate_prefix_config cfg) {
       },
       [this] {
           auto ptr = _segs.front();
-          _segs.pop_front();
-          _probe->add_bytes_prefix_truncated(ptr->file_size());
-          return remove_segment_permanently(ptr, "remove_prefix_full_segments");
+          /**
+           * If segment has outstanding locks wait for it to be unlocked before
+           * prefixing truncating it, this way the prefix truncation will
+           * not overlap with appends.
+           */
+          return ptr->write_lock().then(
+            [this, ptr](ss::rwlock::holder lock_holder) {
```

Comment:
Makes sense to me, though worth noting that it seems like the "Segment has outstanding locks. Might take a while to close:{}" line in remove_segment_permanently will always be logged now

Reply:
this is true, i will remove that line